### ISSUE TITLE
GitButler not using includeIf configurations

### ISSUE THREAD

**Issue by Kristijan** [gitbutlerapp/gitbutler#5420]
### Version

0.13.8

### Operating System

macOS

### Distribution Method

dmg (Apple Silicon)

### Describe the issue

GitButler doesn't appear to use git configurations that are dynamically sourced in via an `includeIf`. I don't have the following configuration set it my default `~/.gitconfig`, rather, they're sourced in depending on the remote that I'm working on.

- `user.name`
- `user.email`
- `user.signingkey`
- `gpg.ssh.allowedsignersfile`

My global `~/.gitconfig` has entries like this, which then sets the respective configuration items.

```
includeif.hasconfig:remote.*.url:git@github.com:*/**.path=~/.dotfiles/git/gitconfig-personal
includeif.hasconfig:remote.*.url:git@ghe.work:*/**.path=~/.dotfiles/git/gitconfig-work
```

I get the following error when trying to commit with the above configuration.

<img width="481" alt="image" src="https://github.com/user-attachments/assets/8d9c179c-36a6-4c5d-aec7-5adb8ad080de">

However, if I do set those variables in either the global or repository configurations, then it works fine.

### How to reproduce

As a minimum, remove the following from your global configuration, and set them via an `includeIf` directive.

- `user.name`
- `user.email`

Use GitButler to make a commit to reproduce the error.

### Expected behavior

GitButler should use any values sourced in via an `includeIf`, just like `git` currently does. It allows the user to set different names, emails, signatures, etc... depending on the `includeIf` test conditions.

**Comment by Byron** [comment]
Thanks a lot for reporting!

`hasconfig` is a relatively 'recent' feature that isn't implemented in `gitoxide` yet. It's also not implemented in `git2` so the only way to respect this today is to use `git` itself to read this configuration. It's probably better to wait until it's available, it's on my list now as this kind of compatibility is very important to me.

**Comment by Byron** [comment]
This is now implemented in `gitoxide` and will soon [land here](https://github.com/gitbutlerapp/gitbutler/pull/5478).

If everything goes well, tomorrow or in a couple of days at the latest, the feature can be tried [in a nightly build](https://app.gitbutler.com/downloads).

**Comment by Kristijan** [comment]
Amazing! Happy to test once it lands in the nightly build.

**Comment by Kristijan** [comment]
Hey @Byron, are you only supporting certain configuration options that are being included with `includeIf`? From my quick limited testing, it appears that `user.name` and `user.email` work, but `user.signingkey` does not (i.e. it doesn't auto-populate the signing key in the project settings for signed commits)?

On the topic of signing keys, is it possible to have this turned on at a global level? I sign all my commits in all my projects, but right now, I need to explicitly turn it on per-project when I add a local repository. Would be nice if I could set a default in the GitButler settings to have signed commits turned on/off, and then manage, should I need to, via the individual project settings.

**Comment by Byron** [comment]
Thanks a lot for sharing your insights!

Indeed, I forgot about all the other configuration keys listed here and only focussed on `user.email` and `user.name`, which are some of the few settings that are actually read with `gitoxide`.

`user.signingKey` for example is read here…

https://github.com/gitbutlerapp/gitbutler/blob/1031de1659bbfea7d72902ef7efcdc0c2de12289/crates/gitbutler-repo/src/repository_ext.rs#L355

…which uses `git2` still, so the `includeIf` support is very limited.

In order to fix this, one would have to use `gitoxide` to read all configuration, or more specifically here, offer a way to create signed commits right away.

**Comment by Kristijan** [comment]
Manually adding `user.signingKey` per-project is just a friction point at this stage for me that I'm willing to live with while `gitoxide` matures. I work in a large organisation using GItHub Enterprise, so it's not uncommon that I'm flipping between 10 different repositories in a typical work week. I'm trying to get GitButler working for me so I can get it to a point of adding a repository and it being usable without having to dive into the settings.

The other option that I set using `includeIf` is `allowedSignersFile`. I have a script that uses the GitHub API to download all the public keys from contributors, and puts them into a file that the option is pointed at. This way, I can use `git show --show-signature` to verify signed commits. Would be nice to see the verified check-mark in GitButler when looking at other users commits. Maybe a future roadmap feature 😄

**Comment by Byron** [comment]
Thanks a lot for holding out for us to catchup :), it's much appreciated!

> This way, I can use `git show --show-signature` to verify signed commits. Would be nice to see the verified check-mark in GitButler when looking at other users commits. Maybe a future roadmap feature 😄

I don't think there is an issue for that yet and I welcome you to set it up :).

**Comment by prom3theu5** [comment]
Currently have an issue getting to work with 1password, and multiple git accounts.
I might be wrong, and its not supported yet - but I was under the impression that if I used the git executable on the path option, then it'd use my current git config?

My current git config conditionally applies .gitconfig merge in based on the url of the remote origin of the repository that is open.
I've tried two ways - using `hasconfig:remote`, and `gitdir:` 

I realise that hasconfig is relatively new, and probably related to this, however gitdir has been around ages and would have expected that to work?
For example

`~/.gitconfig (hasconfig)`

```toml
[includeIf "hasconfig:remote.*.url:**/*github.com*/ipromknight/**"]
  path = ~/.gitconfigs/.gitconfig-promknight
[includeIf "hasconfig:remote.*.url:**/*github.com*/prom3theu5/**"]
  path = ~/.gitconfigs/.gitconfig-prom3theu5
[init]
    defaultBranch = main
```

`~/.gitconfig (gitdir)`

```toml
[includeIf "gitdir:~/git/work/**"]
  path = ~/.gitconfigs/.gitconfig-prom3theu5
[includeIf "gitdir:~/git/personal/**"]
  path = ~/.gitconfigs/.gitconfig-promknight
[init]
    defaultBranch = main
```

`~/.gitconfigs/.gitconfig-prom3theu5`

```toml
[user]
  signingkey = ssh-ed25519 ...
  name = "Prom3theu5"
  email = "..."

[gpg]
  format = ssh
  program = "/Applications/1Password.app/Contents/MacOS/op-ssh-sign"

[commit]
  gpgsign = true
```

`~/.gitconfigs/.gitconfig-promknight`

```toml
[user]
  signingkey = ssh-ed25519 ...
  name = "iPromKnight"
  email = "..."

[gpg]
  format = ssh
  program = "/Applications/1Password.app/Contents/MacOS/op-ssh-sign"

[commit]
  gpgsign = true
```

It always seems to use the Prom3theu5 identity, unless its not supported currently?
Really want to start using Gitbutler, but i work on personal and work projects daily lol
Is there a better way to achieve what I need?

**Comment by Byron** [comment]
Thanks a lot for sharing!

Even though this issue has been neglected and is missing an update, all kinds of conditional includes *should* already work, both with `hasconfig` as well as with `gitdir`.

I will take a closer look at that soon.

**Comment by Byron** [comment]
I did test it and it seemed not to work initially.

However, it turned out that GitButler currently writes changes to the signing options from the user interface directly into the `.git/config` file of the repository at hand, which usually serves as final override.

When that was removed, it worked right away for the `gitdir` version.

When trying the `hasconfig` version, I ran into the behaviour that the remote-url matching is case-sensitive. But once the case was adjusted, it saw the desired value in the UI as well, all using Version 0.14.6 (20250121.112215).

Could you check if `git config -l --show-origin` from the repository in question also shows the expected overrides?
From all I can tell it's working as expected.

Thank you.

**Comment by DerAlbertCom** [comment]
Wanted just to chime in, have more or less the same problem.

I was getting **Need committer to be configured when creating a new commit** until I set `user.email` in my global `.gitconfig`. The ones with `includeif` seem not to be picked up.

Version 0.14.13, Windows 11

**Comment by prom3theu5** [comment]
> I did test it and it seemed not to work initially.
> 
> However, it turned out that GitButler currently writes changes to the signing options from the user interface directly into the `.git/config` file of the repository at hand, which usually serves as final override.
> 
> When that was removed, it worked right away for the `gitdir` version.
> 
> When trying the `hasconfig` version, I ran into the behaviour that the remote-url matching is case-sensitive. But once the case was adjusted, it saw the desired value in the UI as well, all using Version 0.14.6 (20250121.112215).
> 
> Could you check if `git config -l --show-origin` from the repository in question also shows the expected overrides? From all I can tell it's working as expected.
> 
> Thank you.

Thanks Byron I’ll give it a test tonight. Sorry missed the replies

**Comment by prom3theu5** [comment]
This does work with ssh - thats cool thanks
I have a further issue now though... If i need multiple ssh identities (i use 1password as my agent), I have to configure ssh config like so

```bash
Host *
    IdentityAgent ~/.1password/agent.sock

Host prom3theu5
	HostName github.com
	User git
	IdentityFile ~/.ssh/prom3theu5_github.pub
	IdentitiesOnly yes

Host promknight
	HostName github.com
	User git
	IdentityFile ~/.ssh/promknight_github.pub
	IdentitiesOnly yes
```

However - if i do this, gitbutler seems to no longer be able to identify that its connected to a github repo. I get no support for anything other than sync and push
So no pull requests etc


I guess its matching the host explicitly to github.com, instead of HostName?

**Comment by Byron** [comment]
This would be an enhancement for another issue. I think the trick here could be that GitButler associates a name with each login which can then be picked using Git configuration. That would make it possible to select which login to use. CC @krlvi just to float the idea.

**Comment by Byron** [comment]
Closing, as I think the original issue is solved.
Thanks everyone for your help!
